### ISSUE TITLE
Queue corruption when omrelp is blocked and rsyslog force-terminated

### ISSUE THREAD

**Issue by StrongestNumber9** [rsyslog/rsyslog#2742]
### Expected behavior
Queue file works as expected

### Actual behavior
Seems like events in queuefile gets corrupted / duplicated and eventually the whole queuefile vanishes. Ruleset getting malformed is one easily visible example;

```sh
+pszRuleset:1:6:▒▒=▒:
```

### Steps to reproduce the behavior
Bind imrelp to ruleset which has omrelp action with disk queue configured, and omrelp target cannot be reached. 

### Environment
```sh
rsyslogd -v
rsyslogd 8.32.0, compiled with:
  PLATFORM:       x86_64-unknown-linux-gnu
  PLATFORM (lsb_release -d):
  FEATURE_REGEXP:       Yes
  GSSAPI Kerberos 5 support:    No
  FEATURE_DEBUG (debug build, slow code): Yes
  32bit Atomic operations supported: Yes
  64bit Atomic operations supported: Yes
  memory allocator:     system default
  Runtime Instrumentation (slow code): No
  uuid support:       Yes
  systemd support:      No
  Number of Bits in RainerScript integers: 64

See http://www.rsyslog.com for more information.
```

```sh
CentOS Linux release 7.4.1708 (Core)
```

```sh
Linux snip 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by rgerhards** [comment]
an important triggering conditions seems to be that rsyslog is force-terminated. So required order of events:

1. rsyslog starts up
    a. omrelp blocks
    b. rsyslog is force-terminated (e.g. by OS script)
2. rsyslog starts up again
    only know the queue corruptions seems visible.

**Comment by rgerhards** [comment]
@StrongestNumber9 have you ever seen this problem on any run but the second? I mean

- initial startup
- initial force-termination
- 2nd startup
  mangled messages here
- 2nd force-termination
- n+1 startup
  no grabled messages
- ...

I mean if you have seen it in any of the n+1 steps?

**Comment by StrongestNumber9** [comment]
If I remember correctly, the broken ruleset error appeared on just one startup. On further restarts I noticed my queuefiles started having duplicated events, some with broken ruleset:

originally 10 events
restart, events from 10 -> 18 (+8 events)
18 -> 24 (+6)
24 -> 28 (+4)
28 -> 30 (+2)
30 -> 31 (+1)
31 -> statefiles disappeared on shutdown

**Comment by rgerhards** [comment]
thx. I still cannot reproduce with a proper testbench test, unfortunately. But the core problem seems to be that rsyslog gets force-killed. That is possibly caused by the retry handler not sufficiently fast giving up on termination request. But as I cannot reproduce it in a productive debug environment, it's currently very hard to get hold of the actual trigger.

**Comment by rgerhards** [comment]
@StrongestNumber9 Can you send me, via private mail, the patched rsyslog source you use for building. I need to lookup lines and I have the feeling that some line numbers are different from 8.32.0 official.

**Comment by StrongestNumber9** [comment]
@kortemik ping --^

**Comment by rgerhards** [comment]
ok, I am getting somewhat closer. Current idea is that

1. omrelp thread is cancelled because it cannot finish (that is OK)
1. ruleset gets destructed
1. cancel cleanup handler from 1. is now executed
    --> this leaves the ruleset name undefined

I will try to craft a test based on that idea, but it is probably getting tricky getting the exact timing. I need to think what we can do in this case, even if I cannot force the condition inside a test. @StrongestNumber9 could you test a patch in your environment if I remain unable to craft a usable test?

**Comment by rgerhards** [comment]
I forgot: the trigger is not that rsyslog is force-terminated (via signal 9), but rather that the output thread is force-terminated via pthread_cancel.

**Comment by rgerhards** [comment]
@StrongestNumber9 @kortemik looks like I do not necessarily need the patched source at the moment

**Comment by kortemik** [comment]
Hi,

I think I've already also sent the patched source via the drive link as
well :)

-MIkko

**Comment by rgerhards** [comment]
The trigger actually seems to be that a ruleset r1 bound to imrelp calls into another ruleset  r2 (which does not have a queue). It looks like the r1 gets destructed before r2, but the msg object refers back to r1 and that part of the information needs to be read when r2 is being destructed (at which time r1 is already freed). Unfortunately, I still cannot reproduce this with a proper test.

**Comment by rgerhards** [comment]
This looks related to https://github.com/rsyslog/rsyslog/issues/1122, but in regard to the action/ruleset name. Unfortunately, I still cannot manage to get a proper test. So once I craft a fix, I guess I need @StrongestNumber9 to verify that it really works. Is that OK?

**Comment by kortemik** [comment]
Yea we have capability execute test on-demand. Did the provided container images reproduce the problem in your environment?

**Comment by rgerhards** [comment]
> Did the provided container images reproduce the problem in your environment?

Yes, this did work. So it was good to have it. The downside is that I have very limited ability to debug in that environment (albeit as good as possible under the circumstances, so it is well-crafted). That makes it so hard for me to get to a decent point in debugging and fixing. Also for CI I would really love to have a proper test, but well...

I think I can provide a patch soon and I also think chances are 98% that it will fix the issue.

**Comment by rgerhards** [comment]
I have posted the potential fix as PR #2755 - it would be great if you could try it out and report back. Please also let valgrind run. The patch is against master, but should apply to your version of 8.32.0 as well.
